ecu/stats.go

```diff
@@ -32,133 +32,124 @@ import (
 	log "github.com/golang/glog"
 )
 
-// publisher implements an interface for a statistics publisher.
-type publisher interface {
-	update(s *stats)
+// Publisher is an interface for a statistics publisher. Publishers are
+// notified of updates to the statistics with the Update method.
+type Publisher interface {
+	Update(s *Stats)
 }
 
-// stats contains the statistics collected from the Seesaw Engine.
-type stats struct {
-	lock sync.RWMutex
-
-	lastUpdate  time.Time
-	lastSuccess time.Time
-
-	seesaw.ClusterStatus
-	seesaw.ConfigStatus
-	seesaw.HAStatus
-	neighbors []*quagga.Neighbor
-	vlans     []*seesaw.VLAN
-	vservers  map[string]*seesaw.Vserver
+// Stats contains the statistics collected from the Seesaw Engine.
+type Stats struct {
+	LastUpdate  time.Time
+	LastSuccess time.Time
+
+	ClusterStatus seesaw.ClusterStatus
+	ConfigStatus  seesaw.ConfigStatus
+	HAStatus      seesaw.HAStatus
+	Neighbors     []*quagga.Neighbor
+	VLANs         []*seesaw.VLAN
+	Vservers      map[string]*seesaw.Vserver
 }
 
 // ecuStats contains that data needed to the ECU stats collector.
 type ecuStats struct {
-	ecu        *ECU
-	publishers []publisher
-	stats      *stats
+	ecu *ECU
+
+	publishersMu sync.RWMutex
+	publishers   []Publisher
+
+	lastStats *Stats
 }
 
 // newECUStats returns an initialised ecuStats struct.
 func newECUStats(ecu *ECU) *ecuStats {
-	return &ecuStats{
-		ecu: ecu,
-		stats: &stats{
-			lastUpdate:  time.Unix(0, 0),
-			lastSuccess: time.Unix(0, 0),
-		},
-	}
+	return &ecuStats{ecu: ecu}
 }
 
-// notify registers a publisher for update notifications.
-func (e *ecuStats) notify(p publisher) {
-	e.publishers = append(e.publishers, p)
+// notify registers publishers for update notifications.
+func (e *ecuStats) notify(pubs ...Publisher) {
+	e.publishersMu.Lock()
+	defer e.publishersMu.Unlock()
+	e.publishers = append(e.publishers, pubs...)
+}
+
+func (e *ecuStats) runOnce() {
+	log.Info("Updating ECU statistics from Seesaw Engine...")
+	start := time.Now()
+	s, err := e.stats()
+	if err != nil {
+		log.Warningf("Couldn't update statistics: %v", err)
+		s = e.lastStats
+	}
+	s.LastUpdate = start
```

Comment:
do this at the begining with s.LastUpdate = time.Now()?

Looks like it's anyway assigned.

Reply:
It's a timestamp of the start of gathering stats, so it can't be set after the call to stats. Setting on e.lastStats works.

Reply:
I mean you don't need a  local variable "start" to hold now. just assign s.LastUpdate at be begining of the function.

Reply:
I'm confused. s doesn't exist until it is declared (in the line with the call to e.stats), so such an assignment won't work at the start of the function?

Reply:
sorry. I didn't notice stats is newly created. then I find the original version is easier to understand. could you revert back to it?

Reply:
Done.